### PR TITLE
Add a wrapper in gleam to time each gl call.

### PR DESCRIPTION
**Pull request by bholley**
Differential Revision: https://phabricator.services.mozilla.com/D22201

---
This change is [Reviewable](https://reviewable.io/reviews/servo/gleam/187)

### CODE REVIEW

**Comment by bholley** [comment]
@gw3583 Can you merge + release?

File: src/gl.rs [hunk 1]
```diff
@@ -584,6 +597,20 @@ impl<F: 'static + Fn(&Gl, &str, GLenum)> ErrorReactingGl<F> {
     }
 }
 
+/// A wrapper around GL context that times each call and invokes the callback
+/// if the call takes longer than the threshold.
+pub struct ProfilingGl<F> {
+    gl: Rc<Gl>,
```
Comment (nox) [inline src/gl.rs]: Shouldn't that be a type parameter itself? If `ProfilingGl` is used as a trait object that's twice as many virtual calls, right?

Comment (bholley) [reply src/gl.rs]: I'm not sure how that'd work. This needs to be usable interchangeably with other Rc<Gl> references, similar to ErrorReactingGl.

**Comment by asajeffrey** [comment]
I share @nox's concern that this is doubling the number of dynamic function calls.

We might be able to avoid that by something like:
```rust
   pub struct ProfilingGl<F, G: Gl> {
    gl: G,
    ...
```
then adding a method to Gl:
```rust
  fn profile<F: Fn(&str, Duration)>(&self, f: F) -> Rc<Gl> {
     let profiled: ProfilingGl<F, Self> = ...;
     Rc::new(profiled)
  }
```

**Comment by bholley** [comment]
I tried this for a few hours today and wasn't able to make it work. First I got stuck on not being able to make this a default trait impl. I finally gave up and just implemented on each of the concrete types with a macro, but _then_ I got stuck with the inability to downcast Rc<Gl> to Rc<Self> (There is Rc<Any>::downcast, but you can't make an Rc<Any> from Rc<Trait>, only from Rc<ConcreteType>).

So if either of you can come up with a working alternative, please do. Otherwise, please merge the PR (it was already reviewed by @gw3583 ).

**Comment by asajeffrey** [comment]
IRC chat: https://mozilla.logbot.info/servo/20190307#c16061804 and https://mozilla.logbot.info/servo/20190308#c16061834 ending with "All this is a lot of work to save some dynamic method dispatches which probably aren't the bottleneck anyway."

**Pull request merged**